### PR TITLE
Add wolfcrypt test: R/O filesystem const memory pointer

### PR DESCRIPTION
**Pull request by gojimmypi**
# Description

As noted in https://github.com/wolfSSL/wolfssl/issues/6522, the wolfSSL wolfcrypt test currently fails duing a memory access operation on
a read-only embedded filesystem such as Xtensa Linux kernel on the ESP32-S3 when accessing
const var pointers.

This failure does not currently occur during the memory test. This PR fixes that with a new test.

With `WOLFSSL_DEBUG` turned on, the testwolfcrypt will now show the failure during the memory test.

Without this change, the memory test would pass, but the next test (`base64_test()`) would fail:

```
/opt # ./testwolfcrypt
Math:   Multi-Precision: Wolf(SP) word-size=32 bits=3072 sp_int.c
------------------------------------------------------------------------------
 ESP32-S3 Xtensa 1.0a wolfSSL version 5.6.3
----------------------[   16.551666] Caught unhandled exception in 'testwolfcrypt' (pid = 63, pc = 0x42e3c7e2) - should not happen
[   16.551666]  EXCCAUSE is 3
[   16.564861]  3a 37 32 03 00 e0 a0 34 3c f5 90 80 54 e0 60 54 90 95 c0 8a 87 6a 67 40 aa 20 d2 08 00 32 41 30
--------------------------------------------------------
error    test passed!
MEMORY   test passed!
Illegal instruction
```

Here's an example of the new failure message during the memory test:

```
~ # cd /opt
/opt # ./testwolfcrypt
Math:   Multi-Precision: Wolf(SP) word-size=32 bits=3072 sp_int.c
------------------------------------------------------------------------------
 wolfSSL version 5.6.3
------------------------[   16.738576] Caught unhandled exception in 'testwolfcrypt' (pid = 63, pc = 0x42e2a61a) - should not happen
[   16.738576]  EXCCAUSE is 3
[   16.751910]  20 00 a8 01 da 99 98 09 aa 99 92 09 00 c0 20 00 92 41 08 86 05 00 a1 7e df da 9c 98 09 da aa b8
------------------------------------------------------
error    test passed!
Testing const byte ptr reference...
Illegal instruction
```

The fix for this is to add `-mforce-l32` to the wolfSSL `./configure` command. Thanks @jcmvbkbc for the [suggestion](https://twitter.com/jcmvbkbc/status/1670159419089494016?s=20). 

Here's an example of the memory test with the proper compiler flags and `WOLFSSL_DEBUG` turned on:

```
Running sysctl: OK
seedrng: can't create directory '/var/lib/seedrng': Read-only file system
Starting network: OK

Welcome to Buildroot
buildroot login: root
~ # cd /opt
/opt # ./testwolfcrypt
Math:   Multi-Precision: Wolf(SP) word-size=32 bits=3072 sp_int.c
------------------------------------------------------------------------------
 wolfSSL version 5.6.3
------------------------------------------------------------------------------
error    test passed!
Testing const byte ptr reference...
const byte ptr test success
MEMORY   test passed!
```

For reference, here's the entire wolfSSL `./configure` command for Xtensa Linux Kernel on the ESP32-S3:

```
./configure \
  --host=xtensa-esp32s3-linux-uclibcfdpic \
  CC=xtensa-esp32s3-linux-uclibcfdpic-gcc \
  AR=xtensa-esp32s3-linux-uclibcfdpic-ar \
  STRIP=xtensa-esp32s3-linux-uclibcfdpic-strip \
  RANLIB=xtensa-esp32s3-linux-uclibcfdpic-ranlib \
  --prefix=/mnt/s3linux/build-xtensa-2023.02-fdpic-esp32s3/target/opt \
  CFLAGS="-mfdpic -mforce-l32  -DNO_WRITEV -DWOLFSSL_USER_SETTINGS -DDEBUG_WOLFSSL -DTFM_NO_ASM -DALT_ECC_SIZE" \
  --disable-filesystem --disable-all \
  --disable-shared --enable-psk --enable-sha --disable-fpecc --disable-ecc --disable-rsa --disable-sp-asm --disable-sha384 --disable-sha512 --disable-sha224 --disable-md5

```
* edit:  See also this informative presentation: [ARM FDPIC Toolchain and ABI](https://static.linaro.org/connect/sfo15/Presentations/09-24-Thursday/SFO15-406-%20ARM%20FDPIC%20Toolchains.pdf). Much (or all?) of the FDPIC information also applies to the Xtensa environment. The section about "_Pointers to global data are indirectly referenced by the Global Offset Table (GOT)_" is particularly relevant to the changes in the PR.

Fixes zd#  n/a

# Testing

How did you test?

Ran wolfCrypt [test](https://github.com/gojimmypi/wolfssl/tree/wc-memory-test/wolfcrypt/test) in:

* ESP-IDF Embedded app
* Xtensa ESP32-S3 Linux app
* Regular compiled linux app.

# Checklist

 - [x] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Comment by gojimmypi** [comment]
The only PRB failure appears to be unrelated to the change:

```
*** Failed to download the package list from https://downloads.openwrt.org/releases/21.02-SNAPSHOT/targets/x86/64/ ...
```

**Comment by bandi13** [comment]
> The only PRB failure appears to be unrelated to the change:
> 
> ```
> *** Failed to download the package list from https://downloads.openwrt.org/releases/21.02-SNAPSHOT/targets/x86/64/ ...
> ```

Looks like this error is happening on other PRs as well. Probably safe to ignore until we get a fix in.

**Review by dgarske — COMMENTED** [review]
Jim, are you looking to ensure that .const globals are actually loaded? This test seems odd to require and need on all platforms. The PR is like 20 lines of actual code and 60 lines of comments. Are you really wanting this PR to be merged into master still?

**Comment by gojimmypi** [comment]
Hi David, thanks for taking a look at this.

> are you looking to ensure that .const globals are actually loaded? 

Not really. The .const globals, to my understanding, are not actually "loaded". As everything is on the same flash: read-only kernel, compiled app, and all... when the binary app runs, the .const values are _not_ loaded into memory, rather they stay on the file system as an extreme method of resource conservation. 

The trick is to tell the compiler this. It was not at all intuitive to me.  I learned this from @jcmvbkbc (see [tweet](https://twitter.com/jcmvbkbc/status/1670225842516144128?s=20)):

* "_The function peek(const byte *) loads a byte from the address passed to it and the function main calls it with the address of a constant. The constant goes into .rodata section, .rodata stays in the FLASH with the rest of .text when the program runs._"

* "_You can't do 1-byte loads from FLASH mapped through the instruction bus, but you can do an aligned 4-byte load and extract the byte from it. That's what this compiler option does._"

* "_And it does so by turning one l8ui into l32i + 5 other instructions which only takes ~5 cycles more, as opposed to taking an exception and fixing it up in the handler, which takes ~120._"

> This test seems odd to require and need on all platforms. 

I agree, however it is not an intrusive test that would cause any harm. I also do not know how to detect at runtime the fact that a compile-time directive was missing. 

It appears this was all caused by the cross-ng build (see [tweet](https://twitter.com/jcmvbkbc/status/1670159419089494016?s=20)):

* "_This option is also used to build all software in the esp32 buildroot._"

> The PR is like 20 lines of actual code and 60 lines of comments.

Agreed this is a bit odd. I typically add a lot of detail to code that is not very intuitive. This is some of the most unintuitive code I've ever written. Given the unintuitive nature of what it is doing, future edits might cause the test to not actually detect the problem.

> Are you really wanting this PR to be merged into master still?

I'm open to suggestion for improvement, but yes. 

Any future developers of a read-only kernel app will greatly appreciate the error occurring in the memory test and not some arbitrary code that accesses const values that required the `-mforce-l32` option.

**Comment by gojimmypi** [comment]
After reconsidering the feedback, I've updated the changes to the new memory test. There are fewer comments and a couple of additional `WOLFSSL_MSG` debug messages that will be helpful to the embedded developer.

I've resynced with upstream master, retested with both `-mforce-l32` to confirm success and a missing `-mforce-l32` to confirm the error is properly manifested. All commits squashed & ready for review.

The test is not very intrusive. I propose letting it run on all platforms and configurations.

File: wolfcrypt/test/test.c [hunk 1]
```diff
@@ -13556,9 +13560,50 @@ static wc_test_ret_t simple_mem_test(int sz)
...
 
+/* If successful, returns the first letter of the byte array `in`.
+**
+** This is a deceptively simple test of a read-only embedded Linux file system.
+** (e.g CFLAGS `-mfdpic` and `-mforce-l32`) When `-mforce-l32` is missing,
+** access to `in` will fail as the const is on the R/O memory-mapped file
+** system. Const is *not* loaded into app memory. See PR 6523.
+**
+** Edit with caution. */
+static wc_test_ret_t const_byte_ptr_test(const byte* in, word32 *outJ)
```
Comment (dgarske) [inline wolfcrypt/test/test.c:13571]: This code is still too complex and too verbose (logging wise). You can run the entire test with a single code line right?

File: wolfcrypt/test/test.c [hunk 2]
```diff
@@ -13556,9 +13560,50 @@ static wc_test_ret_t simple_mem_test(int sz)
...
+{
+    volatile wc_test_ret_t ret;
```
Comment (dgarske) [inline wolfcrypt/test/test.c:13573]: Do you really understand the use of `volatile` here?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfcrypt/test/test.c (hunk 2 above)
Comment (gojimmypi) [reply wolfcrypt/test/test.c:13573]: I think so. I was having problems with the compiler optimizing away things.

File: wolfcrypt/test/test.c (hunk 1 above)
Comment (gojimmypi) [reply wolfcrypt/test/test.c:13571]: I did try pruning it down to something smaller. The compiler seems to optimize it and the `Illegal Instruction` is not caught during this test. I'll see what I can do here.

File: wolfcrypt/test/test.c [hunk 3]
```diff
@@ -13556,9 +13560,50 @@ static wc_test_ret_t simple_mem_test(int sz)
...
+    volatile byte curChar = ' ';
+    volatile word32 j = -1;
+
+    WOLFSSL_MSG("Testing const byte ptr reference...");
```
[the remark this answers is not in the document]
Comment (gojimmypi) [reply wolfcrypt/test/test.c:13577]: Which do you prefer & should that be in this PR?

I could simply wrap some print's in a `#ifdef DEBUG_WOLFSSL` to match the majority, but there's a bigger issue here.

I used [WOLFSSL_MSG](https://github.com/wolfSSL/wolfssl/blob/6b240fa41a2a3c15bd8a816f7a39b0daccb14635/wolfcrypt/src/logging.c#L350) logging for debug messages to not clutter things up under normal circumstances.

My wolfcrypt test prints almost nothing other than "success" / "fail" for each test unless debugging is turned on.

I see that many of the printf's are not regular printf's. For example there's a macro such as the [zepher for printk](https://github.com/wolfSSL/wolfssl/blob/master/wolfcrypt/test/test.c#L127).

I also just learned that all of those are wrapped in a [#ifndef WOLFSSL_LOG_PRINTF](https://github.com/wolfSSL/wolfssl/blob/6b240fa41a2a3c15bd8a816f7a39b0daccb14635/wolfcrypt/test/test.c#L102C9-L102C27)

It's not very intuitive to have a printf that's not necessarily really a printf.

Perhaps there should be at least one printf macro added for the Espressif ESP-IDF & another for this Linux Xtensa to match the local pattern. Do you agree?

What are your feelings on `WOLFSSL_MSG`? Although the minority in this file, I like the fact that it is all uppercase (clearly a macro), and all the platform-specific and logging level can be handled on a global scale. It does lack the flexibility of printf. But that could be addressed. I'd think we should be using `WOLFSSL_MSG` everywhere, no?

[the remark this answers is not in the document]
Comment (dgarske) [reply wolfcrypt/test/test.c:13577]: Use printf. The printf should already be remapped properly for ESP32 at the top.

File: wolfcrypt/test/test.c (hunk 1 above)
Comment (dgarske) [reply wolfcrypt/test/test.c:13571]: Does this not work? `if (const_byte_array[0] != CBPTR_EXPECTED)`

File: wolfcrypt/test/test.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (gojimmypi) [reply wolfcrypt/test/test.c:13577]: Do you mean the [#elif defined(WOLFSSL_ESPIDF)](https://github.com/wolfSSL/wolfssl/blob/master/wolfcrypt/test/test.c#L120C30-L120C30)? That does not apply in this case as the ESP-IDF is not being used and there's no `WOLFSSL_ESPIDF` and no `ESP_LOG()`.

Recall this PR is for compiling to Linux with `-mforce-l32` for a super tiny, read-only file system. 

For example, here's my current `./configure` command. 

```
./configure \
  --disable-shared \
  --host=xtensa-esp32s3-linux-uclibcfdpic \
  CC=xtensa-esp32s3-linux-uclibcfdpic-gcc \
  AR=xtensa-esp32s3-linux-uclibcfdpic-ar \
  STRIP=xtensa-esp32s3-linux-uclibcfdpic-strip \
  RANLIB=xtensa-esp32s3-linux-uclibcfdpic-ranlib \
  --prefix=/mnt/s3linux/wolfssl-prefix \
  --libdir=/mnt/s3linux/wolfssl-prefix/lib \
  --includedir=/mnt/s3linux/wolfssl-prefix/include \
  --oldincludedir=/mnt/s3linux/wolfssl-prefix/include \
  CPPFLAGS="-I./" \
  CFLAGS=" -mfdpic -mforce-l32 \
           -DTIME_T_NOT_64BIT -DNO_WRITEV -DHAVE_PK_CALLBACKS  \
           -DWOLFSSL_USER_IO -DWOLFSSL_USER_SETTINGS -DDEBUG_WOLFSSL  \
           -DTFM_NO_ASM -DALT_ECC_SIZE"
```

Ok if I simply wrap my new printf's with `#if defined(DEBUG_WOLFSSL)` ?

[the remark this answers is not in the document]
Comment (dgarske) [reply wolfcrypt/test/test.c:13577]: That's fine, but please convert most of those to inline code comments. I don't think there is much value in all those printfs.

**Comment by gojimmypi** [comment]
I'll be delayed on resolving this issue, as pulling from upstream master I now see these new cross-compile-time errors in `api.c` code changed recently by @SparkiDev  in [578f56e6](https://github.com/wolfSSL/wolfssl/commit/578f56e60cae50fc363c42d451be55c9423a8e75):

```
  CCLD     examples/asn1/asn1
  CCLD     testsuite/testsuite.test
  CCLD     tests/unit.test
tests/unit_test-api.o: in function `test_wolfSSL_client_server_nofail_memio':
api.c:(.text+0x3fc3c): dangerous relocation: l32r: literal target out of range (try using text-section-literals): (.literal+0x5464)
api.c:(.text+0x3fc59): dangerous relocation: l32r: literal target out of range (try using text-section-literals): (.literal+0x5458)
api.c:(.text+0x3fc7c): dangerous relocation: l32r: literal target out of range (try using text-section-literals): (.literal+0x543c)
[ ... snip,: many, many more  dangerous relocation: l32r ... ]
```

My current `./configure` looks like this:

```
./configure \
  --disable-shared \
  --host=xtensa-esp32s3-linux-uclibcfdpic \
  CC=xtensa-esp32s3-linux-uclibcfdpic-gcc \
  AR=xtensa-esp32s3-linux-uclibcfdpic-ar \
  STRIP=xtensa-esp32s3-linux-uclibcfdpic-strip \
  RANLIB=xtensa-esp32s3-linux-uclibcfdpic-ranlib \
  --prefix=/mnt/s3linux/wolfssl-prefix \
  --libdir=/mnt/s3linux/wolfssl-prefix/lib \
  --includedir=/mnt/s3linux/wolfssl-prefix/include \
  --oldincludedir=/mnt/s3linux/wolfssl-prefix/include \
  CPPFLAGS="-I./" \
  CFLAGS=" -mfdpic -mforce-l32 \
           -DTIME_T_NOT_64BIT -DNO_WRITEV -DHAVE_PK_CALLBACKS  \
           -DWOLFSSL_USER_IO -DWOLFSSL_USER_SETTINGS -DDEBUG_WOLFSSL  \
           -DTFM_NO_ASM -DALT_ECC_SIZE"
```

File: wolfcrypt/test/test.c (hunk 1 above)
Comment (gojimmypi) [reply wolfcrypt/test/test.c:13571]: correct: I've confirmed that does not properly detect the problem. I'll add a comment as such.

File: wolfcrypt/test/test.c (hunk 2 above)
Comment (gojimmypi) [reply wolfcrypt/test/test.c:13573]: I was a bit too gratuitous on the `volatile`; I've removed the one from `wc_test_ret_t ret` but the next one is essential on `volatile word32 j`; without the `volatile` keyword, I've confirmed the `Illegal instruction` error does not occur.

File: wolfcrypt/test/test.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (gojimmypi) [reply wolfcrypt/test/test.c:13577]: Agreed. Although perhaps a valid test, there's probably no point if I cannot actually make it fail. I've updated the code to use `WOLFSSL_DEBUG` conditional `printf`.

**Comment by gojimmypi** [comment]
[Stackoverflow to the rescue](https://stackoverflow.com/questions/19532826/what-does-a-dangerous-relocation-error-mean) regarding `dangerous relocation: l32r`:

image

> Compiling with `-mtext-section-literals` should eliminate the error

Confirmed. Here's my latest wolfSSL `./configure` for cross-compiling to Xtensa Linux on the ESP32:

```
./configure \
  --disable-shared \
  --host=xtensa-esp32s3-linux-uclibcfdpic \
  CC=xtensa-esp32s3-linux-uclibcfdpic-gcc \
  AR=xtensa-esp32s3-linux-uclibcfdpic-ar \
  STRIP=xtensa-esp32s3-linux-uclibcfdpic-strip \
  RANLIB=xtensa-esp32s3-linux-uclibcfdpic-ranlib \
  --prefix=/mnt/s3linux/wolfssl-prefix \
  --libdir=/mnt/s3linux/wolfssl-prefix/lib \
  --includedir=/mnt/s3linux/wolfssl-prefix/include \
  --oldincludedir=/mnt/s3linux/wolfssl-prefix/include \
  CPPFLAGS="-I./" \
  CFLAGS=" -mfdpic -mforce-l32 -mtext-section-literals \
           -DTIME_T_NOT_64BIT -DNO_WRITEV -DHAVE_PK_CALLBACKS  \
           -DWOLFSSL_USER_IO -DWOLFSSL_USER_SETTINGS -DWOLFSSL_DEBUG  \
           -DTFM_NO_ASM -DALT_ECC_SIZE"
```

**Comment by gojimmypi** [comment]
The 2 PRB errors are likely unrelated to the changes here.

**Comment by dgarske** [comment]
Looks good. I want to run this on some embedded targets also. We may need to wrap the test with a way to disable it. 
Retest this please

**Review by dgarske — DISMISSED** [review]
Tests out fine on STM32U5 CortexM. Thanks

**Comment by gojimmypi** [comment]
No changes, but there were recent upstream `test.c` changes in conflict. Merged, resolved & squashed.

**Review by dgarske — APPROVED** [review]

**Pull request merged**